packages/flutter_tools/lib/src/migrate/migrate_config.dart

```diff
@@ -0,0 +1,257 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:yaml/yaml.dart';
+
+import '../base/common.dart';
+import '../base/file_system.dart';
+import '../base/io.dart';
+import '../cache.dart';
+import '../flutter_project_metadata.dart';
+import '../globals.dart' as globals;
+import '../project.dart';
+
+/// Represents one .migrate_config file.
+///
+/// Each platform and the root project directory includes one .migrate_config file.
+/// This file tracks the flutter sdk git hashes of the last successful migration and the
+/// version the project was created with.
+///
+/// Each platform contains its own .migrate_config file because flutter create can be
+/// used to add support for new platforms, so the base create version may not always be the same.
+class MigrateConfig {
+  /// Creates a MigrateConfig by explicitly providing all values.
+  MigrateConfig({
+    required this.platform,
+    this.createRevision,
+    this.baseRevision,
+    required this.unmanagedFiles,
+  });
+
+  /// Creates a MigrateConfig by parsing an existing .migrate_config yaml file.
+  MigrateConfig.fromFile(File file) : unmanagedFiles = <String>[] {
+    final dynamic yamlRoot = loadYaml(file.readAsStringSync());
+    if (!_validate(yamlRoot)) {
+      // Error
+      globals.logger.printError('Invalid migrate config yaml file found at ${file.path}');
+      return;
+    }
+    final YamlMap map = yamlRoot as YamlMap;
+    platform = map['platform'] as String;
+    createRevision = map['createRevision'] as String;
+    baseRevision = map['baseRevision'] as String;
+    if (map['unmanagedFiles'] != null && (map['unmanagedFiles'] as YamlList).isNotEmpty) {
+      unmanagedFiles = List<String>.from((map['unmanagedFiles'] as YamlList).value.cast<String>());
+    } else {
+      unmanagedFiles = <String>[];
+    }
+  }
+
+  /// The name of the config file.
+  static const String kFileName = '.migrate_config';
+
+  /// A mapping of the files that are unmanaged by deafult for each platform.
+  static const Map<String, List<String>> _kDefaultUnmanagedFiles = <String, List<String>>{
+    'root': <String>['lib/main.dart'],
+    'ios': <String>['Runner.xcodeproj/project.pbxproj'],
+  };
+
+  /// The platform this config file is describing.
+  String? platform;
+
+  /// The git revision this platform was initially created with.
+  ///
+  /// Null if the initial create git revision is unknown.
+  String? createRevision;
+
+  /// The git revision this platform was last successfully migrated with.
+  ///
+  /// Null if the project was never migrated or the revision is unknown.
```

Comment:
when would the revision be unknown?

Reply:
Legacy apps may not have this information available as it was created before we tracked it, in which case we use a fallback revision

Reply:
ahh ok.